chain/chain/src/flat_storage_resharder.rs

```diff
@@ -0,0 +1,943 @@
+            left_child_shard,
+            FlatStorageStatus::Resharding(FlatStorageReshardingStatus::CreatingChild),
+        );
+        store_update.set_flat_storage_status(
+            right_child_shard,
+            FlatStorageStatus::Resharding(FlatStorageReshardingStatus::CreatingChild),
+        );
+        store_update.commit()?;
+
+        self.schedule_split_shard(parent_shard, &status, scheduler, controller);
+        Ok(())
+    }
+
+    /// Returns an error if a resharding event is in progress.
+    fn check_no_resharding_in_progress(&self) -> Result<(), StorageError> {
+        // Do not allow multiple resharding events in parallel.
+        if self.resharding_event().is_some() {
+            error!(target: "resharding", "trying to start a new flat storage resharding event while one is already in progress!");
+            Err(StorageError::FlatStorageReshardingAlreadyInProgress)
+        } else {
+            Ok(())
+        }
+    }
+
+    fn set_resharding_event(&self, event: FlatStorageReshardingEventStatus) {
+        *self.resharding_event.lock().unwrap() = Some(event);
+    }
+
+    /// Returns the current in-progress resharding event, if any.
+    pub fn resharding_event(&self) -> Option<FlatStorageReshardingEventStatus> {
+        self.resharding_event.lock().unwrap().clone()
+    }
+
+    /// Schedules a task to split a shard.
+    fn schedule_split_shard(
+        &self,
+        parent_shard: ShardUId,
+        status: &SplittingParentStatus,
+        scheduler: &dyn FlatStorageResharderScheduler,
+        controller: FlatStorageResharderController,
+    ) {
+        let event = FlatStorageReshardingEventStatus::SplitShard(parent_shard, status.clone());
+        self.set_resharding_event(event);
+        info!(target: "resharding", ?parent_shard, ?status,"scheduling flat storage shard split");
+
+        let resharder = self.clone();
+        let task = Box::new(move || split_shard_task(resharder, controller));
+        scheduler.schedule(task);
+    }
+
+    /// Cleans up children shards flat storage's content (status is excluded).
+    fn clean_children_shards(&self, status: &SplittingParentStatus) -> Result<(), Error> {
+        let SplittingParentStatus { left_child_shard, right_child_shard, .. } = status;
+        debug!(target: "resharding", ?left_child_shard, ?right_child_shard, "cleaning up children shards flat storage's content");
+        let mut store_update = self.runtime.store().flat_store().store_update();
+        for child in [left_child_shard, right_child_shard] {
+            store_update.remove_all_deltas(*child);
+            store_update.remove_all_values(*child);
+        }
+        store_update.commit()?;
+        Ok(())
+    }
+}
+
+/// Retrieves the flat head of the given `shard`.
+/// The shard must be in [FlatStorageStatus::Ready] state otherwise this method returns an error.
+fn retrieve_shard_flat_head(shard: ShardUId, store: &FlatStoreAdapter) -> Result<BlockInfo, Error> {
```

Comment:
I think you can be more aggressive here and provide the exact block hash where you want the flat storage to be and assert it has that block hash within reach. It will be more predictable and easier to reason about it once we know what state exactly are we resharding. 
* if flat head is past that block we're screwed - we won't be able to process the first blocks of the children in new shard layout
* if flat head is before that block it should have deltas that can be applied till the desired block

Note that we can't quite force apply the deltas because we'll be resharding on top of a non-final block and if fork happens we would have to revert the deltas.

Reply:
I'll need to think about the presence of deltas in the parent - if we can't force apply deltas does it mean we must split/copy them to the children?

Reply:
We had this solved in resharding v2 -> I think it was along the lines of chaining the head iterator with the deltas iterator and then you just copy paste everything as goes. The only gotcha is that delta may delete something so need to support that.